src/main/resources/js/macro-customise.js

```diff
@@ -34,4 +34,110 @@ AJS.bind('init.rte', function () {
     }
   });
 
-});
\ No newline at end of file
+});
+
+AJS.$(document).ready(function () {
+  var pageId = AJS.Meta.get("page-id");
+  var baseUrl = AJS.Confluence.getBaseUrl();
+  // Get all shadow elements from page
+  // return a dom node list
+  function getAllShadowElements() {
+    var elements = [];
+    AJS.$('sequence-diagram').each(function () {
+      if (this.shadowRoot.querySelector(".sequence-diagram-container")) {
+        elements.push(this.shadowRoot.querySelector(".sequence-diagram-container"))
+      }
+    });
+    return elements;
+  }
+  // Get all Macro nodes and its dsl
+  // return dsl text and convertToBlob promise
+  function getConvertPromiseAndDsl() {
+    var doms = getAllShadowElements();
+    var objects = [];
+    if (doms.length > 0) {
+      var dslList = [];
+      AJS.$.each(doms, function (index, dom) {
+        var dslEls = AJS.$(dom).find('.zenuml-dsl');
+        if (dslEls.length > 0) {
+          var dslText = dslEls.text();
+          // remove duplicate when upload mutiple same uml at the same time
+          if (dslList.indexOf(dslText) < 0) {
+            objects.push({
+              dsl: dslText,
+              cpro: domtoimage.toBlob(dom, { bgcolor: 'white' })
+            });
+            dslList.push(dslText);
+          }
+        }
+      })
+    }
+    return objects;
+  }
+  // Get all attachments based on page id 
+  // return a ajax promise
+  function getAttachments() {
+    return AJS.$.ajax({
+      url: baseUrl + '/rest/api/content/' + pageId + '/child/attachment?expand=version',
+      type: 'GET'
+    });
+  }
+  // Create and Upload a new attchment 
+  // return a ajax promise
+  function uploadAttachment(blob, dsl) {
+    var hash = md5(dsl);
+    var file = new File([blob], "zenuml-" + hash, { type: 'image/png' });
+    var fd = new FormData();
+    fd.append("file", file);
+    fd.append('comment', hash);
+    fd.append('minorEdit', "true");
+
+    return AJS.$.ajax({
+      url: baseUrl + '/rest/api/content/' + pageId + '/child/attachment',
+      type: 'POST',
+      beforeSend: function (request) {
+        request.setRequestHeader("X-Atlassian-Token", "nocheck");
+      },
+      processData: false,
+      contentType: false,
+      data: fd
+    });
+  }
+  // Validate a list of promise
+  // return Promise.all
+  function resolveAllPromises(promiseList) {
+    if (!Array.isArray(promiseList)) {
+      console.error("Wrong Promise Array");
+    }
+    return Promise.all(promiseList);
+  }
+
+  setTimeout(function () {
+    var dslPromiseList = getConvertPromiseAndDsl();
+    var promises = [];
+    var dsls = [];
+    if (dslPromiseList.length > 0) {
+      AJS.$.each(dslPromiseList, function (index, item) {
+        promises.push(item.cpro);
```

Comment:
`cpro: domtoimage.toBlob(dom, { bgcolor: 'white' }` item.cpro does not look like a promise.

Reply:
It is a promise based on domtoimage doc https://github.com/tsayen/dom-to-image, I need to give it a proper name